torch_np/_ndarray.py

```diff
@@ -343,67 +399,80 @@ def __setitem__(self, index, value):
         return self._tensor.__setitem__(index, value)
 
 
-def asarray(a, dtype=None, order=None, *, like=None):
-    _util.subok_not_ok(like)
-    if order is not None:
+# This is the ideally the only place which talks to ndarray directly.
+# The rest goes through asarray (preferred) or array.
+
+def array(object, dtype=None, *, copy=True, order='K', subok=False, ndmin=0,
+          like=None):
+    _util.subok_not_ok(like, subok)
+    if order != 'K':
         raise NotImplementedError
 
-    if isinstance(a, ndarray):
-        if dtype is not None and dtype != a.dtype:
-            a = a.astype(dtype)
-        return a
+    # a happy path
+    if isinstance(object, ndarray):
+        if copy is False and dtype is None and ndmin <= object.ndim:
+            return object
 
-    if isinstance(a, (list, tuple)):
-        # handle lists of ndarrays, [1, [2, 3], ndarray(4)] etc
+    # lists of ndarrays: [1, [2, 3], ndarray(4)] convert to lists of lists
+    if isinstance(object, (list, tuple)):   
         a1 = []
-        for elem in a:
+        for elem in object:
             if isinstance(elem, ndarray):
                 a1.append(elem.get().tolist())
             else:
                 a1.append(elem)
+        object = a1
+
+    # get the tensor from "object"
+    if isinstance(object, ndarray):
+        tensor = object._tensor
+        base = object
+    elif isinstance(object, torch.Tensor):
+        tensor = object
+        base = None
     else:
-        a1 = a
+        tensor = torch.as_tensor(object)
+        base = None
 
-    torch_dtype = _dtypes.torch_dtype_from(dtype)
+        # At this point, `tensor.dtype` is the pytorch default. Our default may
+        # differ, so need to typecast. However, we cannot just do `tensor.to`,
+        # because if our desired dtype is wider then pytorch's, `tensor`
+        # may have lost precision:
 
-    # This and array(...) are the only places which talk to ndarray directly.
-    # The rest goes through asarray (preferred) or array.
-    out = ndarray()
-    tt = torch.as_tensor(a1, dtype=torch_dtype)
-    out._tensor = tt
-    return out
+        # int(torch.as_tensor(1e12)) - 1e12 equals -4096 (try it!)
 
+        # Therefore, we treat `tensor.dtype` as a hint, and convert the
+        # original object *again*, this time with an explicit dtype.
+        dtyp = _dtypes.dtype_from_torch(tensor.dtype)
+        default = _dtypes.get_default_dtype_for(dtyp)
+        torch_dtype = _dtypes.torch_dtype_from(default)
```

Comment:
we may want to overload `get_default_dtype_for` to also accept PyTorch dtypes to simplify this code.

Reply:
Also, I get the feeling that if PyTorch defaults are the same as those in NumPy (via `set_default...`) simply doing `as_tensor` should do the trick. Is this possible?
Otherwise, as you mention above, we should make sure we just call `as_tensor`  from here, as otherwise we'll inadvertently get incorrect results.

Reply:
Let's maybe not rely on this, not just yet at least. When we have a more complete coverage, let's experiment with how much we can peel out.